### PR TITLE
faster generate_square_subsequent_mask in nn.Transformer

### PR DESCRIPTION
**Pull request by bhosmer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60631 faster generate_square_subsequent_mask in nn.Transformer**
* #60593 configurable pre/post LayerNorm in nn.Transformer

Per #48360, speed up `Transformer.generate_square_subsequent_mask`. New impl is informally ~5x faster, though absolute difference is probably small.

PR includes Python and C++ versions as well as a couple of places where the previous impl had been copied around.

Differential Revision: [D29356673](https://our.internmc.facebook.com/intern/diff/D29356673)

### CODE REVIEW

**Comment by bhosmer** [comment]
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29356673).

**Review by jbschlosser — APPROVED** [review]
Looks awesome! Thanks for the update :)

**Comment by bhosmer** [comment]
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29356673).

**Comment by bhosmer** [comment]
@bhosmer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29356673).

File: torch/csrc/api/src/nn/modules/transformer.cpp [hunk 1]
```diff
@@ -422,21 +422,17 @@ Tensor TransformerImpl::generate_square_subsequent_mask(int64_t sz) {
   TORCH_CHECK(sz >= 0,
     "Input size must be non-negative to genearte a valid square subsequent mask, but got ", sz);
 
-  Tensor mask = (torch::triu(torch::ones({sz, sz})) == 1).transpose(0, 1).to(torch::kFloat32);
-
   // check IEEE754 support here since -inf is not guaranteed to be valid on non IEEE754 platform
   if (std::numeric_limits<float>::is_iec559) {
-    mask = mask.masked_fill(mask == 0, -std::numeric_limits<float>::infinity()).masked_fill(mask == 1, 0.f);
+    return torch::triu(torch::full({sz, sz}, -std::numeric_limits<float>::infinity()), 1);
```
Comment (vadimkantorov) [inline torch/csrc/api/src/nn/modules/transformer.cpp:427]: btw if this is a lot faster than expand(), should we file a bug? in theory expand-version allocates less and just needs a single value to be in cache

**Pull request closed without merging**